README.md
# hellolan
Intuitive port scanning! This module is just a couple functions that wrap nmap and are exposed as command line tools.

Basically, I made this because I often need to scan for raspberry pi's on my local network and was always frustrated about remembering the correct commands. Then I needed to do some automated scanning in Python, so I just wrapped it all up in a few python functions with a cli.

## Install

```bash
# mac
brew install nmap
# linux
apt-get install nmap

pip install hellolan
```

## Usage

```bash
# by default, we scan ports 21-443
hellolan scan

# there are some presets too
hellolan ssh # --port=22
hellolan web # --port=80,443
```
Example output:
```
$ hellolan ssh
hostname           ip             ports
-----------------  -------------  -------
                   192.168.1.127  [22]
node-b827e9j315hf  192.168.1.236  [22]
-
Took 3.3 seconds
```

If you run with `sudo`, you can also see the mac address and vendor.

```
$ sudo hellolan ssh --watch
hostname                ip             ports    mac                vendor
----------------------  -------------  -------  -----------------  ---------------------------

raspberrypi  192.168.1.242  [22]     DC:A6:32:C4:F4:A3  Raspberry Pi Trading
raspberrypi  192.168.1.65   [22]     DC:A6:32:A9:F0:82  Raspberry Pi Trading
             192.168.1.66   [22]     52:D4:F7:18:5F:74
             192.168.1.67   [22]     00:1A:62:03:F3:FF  Data Robotics, Incorporated
Scan 1 finished at Tue Jan 19 23:08:55 2021. took 10.2s. Found 4 hosts.
```


There's more:
```bash
# you can filter the results too
hellolan ssh node- # match a partial hostname
hellolan ssh 'host*' --ignore='badhost-*' # matching glob
hellolan ssh 192.168.1.84 # match an exact ip
# NOTE: lmk if you'd prefer regex over fnmatch

# you can change the net hosts to scan.
# by default it's the local router subnet --net=192.168.1.0/24
# so it's not exclusively for lan ...!! helloevbody!
hellolan scan --net=scanme.nmap.org
hellolan scan --net=198.116.0-255.1-127
hellolan scan --net=216.163.128.20/20

# you can also poll localhost:
# here I'm checking what I have for jupyter lab instances running.
# they autoincrement ports as 8888 + i. I'm assuming I don't have
# more than 7.
hellolan scan --net=localhost --port=8888-8893
# Outputs:
#     hostname    ip         ports
#     ----------  ---------  ------------
#     localhost   127.0.0.1  [8888, 8889]
#     -
#     Took 11.4 seconds
```

You can print it out in more parseable formats
```bash
$ hellolan ssh -ip
# 192.168.1.127
# 192.168.1.236
$ hellolan ssh -ip -json
# ["192.168.1.127", "192.168.1.236"]
$ hellolan ssh -json
# [{"hostname": "", "ip": "192.168.1.127", "ports": [22]}, {"hostname": "raspberrypi", "ip": "192.168.1.236", "ports": [22]}]
```

#### Now you can use the command to directly ssh into a device !!!
How it works - it will split by the '@' symbol, poll for a device who's host matches 'abc' and replaces 'abc' with the found ip address.
```bash
hellolan ssh- user@abc
```
So for example:
```bash
mbp $ hellolan ssh- user@abc
#
# Multiple hosts found:
#     hostname  ip             ports
# --  --------  -------------  -------
#  0  abcdejkl  192.168.1.127  [22]
#  1  abcdefgh  192.168.1.236  [22]
#
# Which host to use? [0]: >> 1
# Using host: abcdefgh
#
# ---------------------
# Starting SSH Session: $ ssh sonyc@192.168.1.236
# ---------------------
#
# Debian GNU/Linux comes with ABSOLUTELY NO WARRANTY, to the extent
# permitted by applicable law.
# Last login: Wed Apr 15 18:40:32 2020 from 192.168.1.214
user@abcdefgh:~ $ hostname
abcdefgh
user@abcdefgh:~ $ exit
logout
Connection to 192.168.1.236 closed.
#
# -------------------
# Ended SSH Session. (ssh sonyc@192.168.1.236)
# -------------------
#
mbp $
```

#### It's also importable!

```python
import hellolan

for d in hellolan.scan(port='21-22'):
    print('{hostname} ({ip}) - {ports}'.format(**d))
```

## Notes
 - this uses the default `nmap.PortScanner().scan(host, ports=ports)` setup. I'm not sure I'm knowledgeable enough to pick better defaults, but there may be ways to speed it up.

 - Update: after playing around with settings, there are indeed ways of speeding it up.
    - I made version checking optional. nmap.PortScanner.scan does version checking by default (`-sV`). I removed that and now you can re-enable it using `--services`. I haven't seen any lost information and the queries are now waaaay faster.
    - if you are trying to speed up `--services`, you can specify version intensity (`--intensity 0-9`) with `0` using fewer checks (faster) and `9` using the most (comprehensive). I noticed significant decrease in runtime when using `--services` with `--intensity 0`
    - You can specify how many of the top ports to use (default is `--top 200`), fewer is faster, (nmap default is `1000`).


CHANGELOG.md
# Changelog

## 0.0.12
 - added CHANGELOG.md
 - added `hellolan me` which returns the current local IP address (matches closest to `192.168.1.`)
 - will return all or a subset of ips if positional arguments or an `--all` flag is passed
 - `hellolan hostname` will now return the current machine's hostname in the absence of a specified IP address
 - added `ifcfg` as a dependency (for determining my lan IP)
 - make assumption that time between scan results is instantaneous compared to rendering (fixes render glitch)

## 0.0.11
 - add natural sorting to cli table
 - removed `force` flag (`nmap -R`) because I saw no improvement over hostname resolution

## 0.0.10
 - added mac and vendor to the table if they are available (if it was run with sudo)

## 0.0.9
 - fixed decorator bug in cli (don't use 0.0.8 directly)

## 0.0.8
 - refactor CLI so that any of the data views are "watch"-able (`--watch`)

## 0.0.7
 - add `--watch` flag to cli table. will perform multiple back to back scans and update the table
 - added `reprint` as a dependency for stdout refreshes
 - moved ssh code into separate file
 - added `hellolan hostname <ip>` command

## 0.0.6
 - add `hellossh` command (shortcut for `hellolan ssh- ...`)

## 0.0.5
 - update dependencies


hellolan/ssh.py
from . import scan


def ssh_into(hostname, *a, user=None, i=None, sshargs=None, port=None, **kw):
    '''ssh into a host by hostname'''
    from subprocess import call

    hostname, user, port = _parse_hostname(hostname, user, port)
    hosts = list(scan(hostname, *a, port=port or 22, **kw))
    host = _resolve_multiple_hosts(hosts, i)

    if not host:
        return

    cmd = 'ssh {} {}'.format(_build_hoststr(host['ip'], user, port), sshargs or '').strip()

    print('''
---------------------
Starting SSH Session: $ {cmd}
---------------------
'''.format(cmd=cmd))
    call(cmd, shell=True)
    print('''
-------------------
Ended SSH Session. ({cmd})
-------------------
'''.format(cmd=cmd))




def _parse_hostname(hostname, user=None, port=None):
    if not user:
        user, hostname = hostname.split('@') if '@' in hostname else (hostname, None)
    if not port:
        hostname, port = hostname.split(':') if ':' in hostname else (hostname, None)
    return hostname, user, port

def _build_hoststr(hostname, user, port):
    if user:
        hostname = '{}@{}'.format(user, hostname)
    if port:
        hostname = '{} -p {}'.format(hostname, port)
    return hostname

def _resolve_multiple_hosts(hosts, i=None):
    print()
    # select which host to connect to
    if not hosts:
        print('No hosts found.')
        return

    if len(hosts) > 1: # by default, ask for multiple
        from tabulate import tabulate
        print('Multiple hosts found:')
        print(tabulate(hosts, headers='keys', showindex=True))
        print()

        if i is None:
            i = int(input('Which host to use? [0]: ').strip() or 0)
    else:
        i = 0

    host = hosts[i or 0]
    print('Using host: {}'.format(host['hostname']))
    return host


hellolan/hellolan.py
import fnmatch
import subprocess
import nmap
import logging

log = logging.getLogger(__name__)

nm = nmap.PortScanner()

def lanscan(net='192.168.1.0/24', port=None, intensity=None,
            nmapargs=None, top=50, services=False, repeat=1, showall=False):
    '''Scan network and ports.

    port='22-433', port=22, port='22,80', port=(80, 443, '2000-2200')
    '''
    if isinstance(port, (tuple, list)):
        port = ','.join(map(str, port))

    nmapargs = (nmapargs,) if isinstance(nmapargs, str) else tuple(nmapargs or ())
    if services:
        nmapargs = ('-sV',)
    if intensity is not None:
        assert 0 <= intensity <= 9
        nmapargs += ('--version-intensity {}'.format(intensity),)
    if top:
        nmapargs += ('--top-ports {}'.format(top),) #('-F',) # -F is not working
    # if force:
    #     nmapargs += ('-R',)

    for i in range(repeat or 1):
        nm.scan(net, ports=str(port) if port else None, arguments=' '.join(nmapargs))
        for ip in nm.all_hosts():
            host = nm[ip]
            ports = [
                port for proto in host.all_protocols()
                for port, status in host[proto].items()
                if status['state'] == 'open'
            ]

            if ports or showall:
                data = {
                    'hostname': nm[ip].hostname() or '',
                    'ip': ip, 'ports': ports,
                    # 'status': host['status']['state'],  # up
                }
                # lookup mac address in case they ran with sudo
                mac = host['addresses'].get('mac')
                vendor = host['vendor'][mac] if mac and mac in host['vendor'] else None
                if mac:
                    data.update({'mac': mac, 'vendor': vendor})
                yield data


def scan(hostname=None, ignore=None, ip=None, n=None, hasname=None, **kw):
    '''Scan devices on your local network. Filter by hostname or ip.
    '''
    # get all devices
    devices = lanscan(**kw)
    if hostname:
        devices = (d for d in devices if matches(d, hostname))
    if ignore:
        devices = (d for d in devices if not matches(d, ignore))
    if ip:
        devices = (d for d in devices if check_ranges([d['ip'], d['ip'].split('.')], ip))
    if hasname:
        devices = (d for d in devices if d['hostname'])
    if n is not None:
        devices = (d for d, i in zip(devices, range(n)))
    # get devices grouped into different categories
    return devices


def hostname(ip=None, port='22-433'):
    if ip is None:
        ip = me()
    nm.scan(ip, str(port) if port else None)
    return nm[ip].hostname()


_SKIP_ME = {'lo0'}
def me(*names, v6=False, all=False):
    import ifcfg
    inet = 'inet6' if v6 else 'inet'
    ips = {k: ifc[inet] for k, ifc in ifcfg.interfaces().items()
           if k not in _SKIP_ME and ifc.get(inet)}
    if names:
        return [ips.get(n) for n in names]
    if all:
        return ips
    return max_common_prefix('192.168.4', *ips.values(), n=2, split='.')


'''

Utils

'''

MATCH_COLS = 'hostname', 'ip'

def matches(d, pat):
    pat = str(pat)
    return any(pat in str(d[c]) or fnmatch.fnmatch(str(d[c]), pat) for c in MATCH_COLS)

def check_ranges(xs, ranges):
    ranges = str(ranges).split(',')
    between = lambda x, xmin, xmax: xmin < x < xmax
    return any(
        (between(x, *r.split('-')) if '-' in r else x == r)
        for x in xs or () for r in ranges
    )


def max_common_prefix(prefix, *values, n=1, split=None):
    if split:
        prefix, values = prefix.split(split), (x.split(split) for x in values)
    match = max((
        (i, x) for i, x in ((n_common_prefix(prefix, val), val) for val in values)
        if i >= n), default=(None, None))[1]
    if match is not None and split:
        match = split.join(match)
    return match

def n_common_prefix(*strs):
    return next(
        (i for i, chs in enumerate(zip(*strs)) if len(set(chs)) > 1),
        min(len(s) for s in strs))


hellolan/__init__.py
from functools import wraps
from .hellolan import *
from .ssh import *

PRESETS = {
    'ssh': 22, 'web': '80,443',
}

def main():
    '''Hellolan CLI'''
    import fire
    fire.Fire({
        'get': partial(getall, n=1),
        'getall': getall,
        'scan': _gentable(scan),
        **{k: _gentable(get_preset(k)) for k in PRESETS},
        'ssh-': ssh_into,
        'hostname': hostname,
        'me': me,
    })

def ssh_main():
    '''SSH Into CLI'''
    import fire
    fire.Fire(ssh_into)


def getall(col, hostname=None, preset=None, *a, **kw):
    return [d[col] for d in get_preset(preset)(hostname, *a, **kw)]


def get_preset(name):
    if name not in PRESETS:
        return scan
    kw = PRESETS[name]
    return partial(scan, **(kw if isinstance(kw, dict) else {'port': kw}))


def _loop(n):
    if n:
        yield from range(n)
    else:
        i = 0
        while True:
            yield i
            i += 1

def _dict_update(prev, new):
    prev, new = prev or {}, new or {}
    return dict(prev, **{k: new[k] or prev.get(k) for k in new})

def _dict_drop(d, *keys):
    return {k: v for k, v in d.items() if k not in keys}

def _gentable(func):
    '''Will print out rows of a table as they are generated by func(*a, **kw).
    I've found out that it's not super necessary because nmap.PortScanner
    yields most things toward the end anyways. So I may end up removing reprint.
    '''
    import time
    import functools
    import itertools
    from tabulate import tabulate

    # def _watch(disp, *a, headers=None, times=None, timer=True, **kw):
    def watchable(disp):
        @functools.wraps(disp)
        def outer(*a, watch=False, headers=None, times=None, timer=True, **kw):
            if not watch:
                # return _watch((lambda data: disp(data, headers)), *a, times=times, timer=timer, **kw)
                result = func(*a, **kw)
                print(disp(list(result), headers=headers))
                return result

            import datetime
            import reprint
            items = {}
            with reprint.output() as out:
                try:
                    out.append('Starting scan...')
                    for i in _loop(times):
                        t0 = time.time()
                        j = 0
                        for j, x in enumerate(func(*a, **kw)):
                            items[x['ip']] = _dict_update(items.get(x['ip']), x)
                        out.change(disp(list(items.values()), headers).splitlines())
                        if timer:
                            out.append('Scan {} finished at {}. took {:.1f}s. Found {} hosts.'.format(
                                i+1, datetime.datetime.now().strftime('%c'),
                                time.time() - t0, len(items)))
                except KeyboardInterrupt:
                    out.change(disp(list(items.values())).splitlines())
            return list(items.values())
        return outer

    @watchable
    def table(items, headers=None, sort='ip'):
        if isinstance(items, dict):
            items = list(items.values())
        if not headers and items and isinstance(items[0], dict):
            headers = 'keys'
        if sort:
            items = sorted(items, key=lambda x: natsort_key(x[sort]))
        return tabulate(items, headers=headers or ())

    @watchable
    def parseable(data, headers=None):
        headers = headers or set().union(*(d.keys() for d in data))
        return '\n'.join([
            '\t'.join(str(d.get(c)) or '' for c in headers)
            for d in data
        ])

    @watchable
    def as_json(data, headers=None):
        import json
        headers = headers or len(data) and data[0].keys() or ()
        data = ([d[headers[0]] for d in data] if len(headers) == 1 else
                [{c: d[c] for c in headers} for d in data])
        return json.dumps(data, indent=4)

    def save_json(out, result, inventory=None):
        import json
        with open(out, 'w') as f:
            if inventory:
                if isinstance(inventory, str):
                    inventory = inventory.split('&')
                inventory = (i.split('=') for i in inventory)
                res = {
                    group: [x['ip'] for x in result if matches(x, pat)]
                    for group, pat in inventory
                }
            else:
                res = {
                    d['ip']: _dict_drop(d, 'ip', 'ports') for d in result
                }
            json.dump(res, f)

    @functools.wraps(func)
    def inner(*a, headers=None, iplist=False, tab=False, timer=True, json=False, out=None, inventory=None, **kw):
        t0 = time.time()
        if iplist:
            headers, tab = (headers or ('ip',)), True
        if json:
            result = as_json(*a, headers=headers, timer=timer, **kw)
        elif tab:
            result = parseable(*a, headers=headers, timer=timer, **kw)
        else:
            result = table(*a, headers=headers, timer=timer, **kw)
        if out:
            save_json(out, result, inventory=inventory)
        if timer and not tab and not json:
            print('-')
            print('Took {:.1f} seconds'.format(time.time() - t0))
    return inner

def partial(func, *a, **kw):
    '''functools.partial doesn't apply wraps ???? wtf ??? it's right there......'''
    return wraps(func)(lambda *ai, **kwi: func(*a, *ai, **kw, **kwi))

import re
def natsort_key(s, _nsre=re.compile('([0-9]+)')):
    return tuple(int(t) if t.isdigit() else t.lower() for t in _nsre.split(s))


if __name__ == '__main__':
    main()
